lnrpc/walletrpc/psbt.go

```diff
@@ -0,0 +1,106 @@
+// +build walletrpc
+
+package walletrpc
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil/psbt"
+	"github.com/btcsuite/btcwallet/wtxmgr"
+	"github.com/lightningnetwork/lnd/lnwallet"
+)
+
+const (
+	defaultMinConf = 1
+	defaultMaxConf = math.MaxInt32
+)
+
+// utxoLock is a type that contains an outpoint of an UTXO and its lock lease
+// information.
+type utxoLock struct {
+	lockID     wtxmgr.LockID
+	outpoint   wire.OutPoint
+	expiration time.Time
+}
+
+// addInputUTXOInfo checks that all partial inputs of a packet are sane witness
+// inputs and are contained in the list of known, non-locked UTXOs given. If
+// successful, all witness UTXO information is added to the partial input
+// section of the PSBT packet.
+func addInputUTXOInfo(packet *psbt.Packet, utxos []*lnwallet.Utxo) error {
+	// To avoid running into index out of bounds problems, make sure the
+	// wire inputs have the same length as the partial inputs. At this point
+	// we expect there to be inputs and outputs too.
+	err := psbt.VerifyInputOutputLen(packet, true, true)
+	if err != nil {
+		return err
+	}
+
+	// Make sure we only have witness inputs. We don't support non-witness
+	// inputs in lnd's wallet. Also make sure the UTXO is known to us and
+	// unspent (and not previously locked).
+	for idx, pIn := range packet.Inputs {
+		if !pIn.IsSane() {
+			return fmt.Errorf("partial input %d is not sane", idx)
+		}
+
+		// We are going to overwrite the witness UTXO information later
+		// anyway. But if we have an invalid value to begin with, don't
+		// continue.
+		if pIn.NonWitnessUtxo != nil {
+			return fmt.Errorf("partial input %d specifies "+
+				"unsupported non-witness input", idx)
+		}
+
+		// Look up the input in the UTXO list.
+		txIn := packet.UnsignedTx.TxIn[idx]
+		var utxo *lnwallet.Utxo
+		for _, u := range utxos {
+			if u.OutPoint == txIn.PreviousOutPoint {
+				utxo = u
+				break
+			}
+		}
+		if utxo == nil {
+			return fmt.Errorf("input %d not found in list of non-"+
+				"locket UTXO", idx)
+		}
+
+		// Everything's OK, we found the UTXO and can add its
```

Comment:
Isn't this already done within the `btcwallet` methods? If not, then why not?

Reply:
You're right, most of this is already done by the wallet. The only thing to check here is that the inputs aren't spent or locked yet.